presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -477,6 +486,24 @@ protected static CacheQuota generateCacheQuota(HiveSplit hiveSplit)
         return Optional.empty();
     }
 
+    private static boolean shouldSkipBucket(HiveTableLayoutHandle hiveLayout, HiveSplit hiveSplit, SplitContext splitContext)
+    {
+        if (!splitContext.getDynamicFilterPredicate().isPresent()
+                || !hiveSplit.getReadBucketNumber().isPresent()
+                || !hiveSplit.getStorage().getBucketProperty().isPresent()) {
+            return false;
+        }
+
+        TupleDomain<ColumnHandle> dynamicFilter = splitContext.getDynamicFilterPredicate().get();
+        Optional<HiveBucketing.HiveBucketFilter> hiveBucketFilter = getHiveBucketFilter(hiveSplit.getStorage().getBucketProperty(), hiveLayout.getDataColumns(), dynamicFilter);
+
+        if (!hiveBucketFilter.isPresent()) {
+            return false;
+        }
+
+        return !hiveBucketFilter.get().getBucketsToKeep().contains(hiveSplit.getReadBucketNumber().getAsInt());
```

Comment:
```java
return hiveBucketFilter.map(filter -> !filter.getBucketsToKeep().contains(hiveSplit.getReadBucketNumber().getAsInt())).orElse(false);
```

Please double check my syntax. I'm not exactly sure I got the negation correctly.